File: EmbeddedNetworkLab.HttpServer/Infrastructure/Services/HttpServerService.cs
using EmbeddedNetworkLab.Core;
using EmbeddedNetworkLab.Core.Models;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Text;
using System.Threading.Tasks;

namespace EmbeddedNetworkLab.Infrastructure.Services
{
	public class HttpServerService : IHttpServerService
	{
		private WebApplication? _app;

		private readonly List<string> _listeningUrls = new();
		public IReadOnlyCollection<string> ListeningUrls => _listeningUrls;

		public bool IsRunning { get; private set; }

		public event EventHandler<string>? RequestReceived;
		public event EventHandler<string>? ServerEventTriggered;
		public event EventHandler<ReceivedVideo>? VideoReceived;
		public event EventHandler<UploadProgress>? UploadProgressChanged;

		public async Task StartAsync(string bindIp, int httpPort, bool httpsEnabled, int httpsPort)
		{
			if (IsRunning) return;

			_listeningUrls.Clear();

			try
			{
				var builder = WebApplication.CreateBuilder();

				// Suppress ASP.NET Core console logs — we forward events ourselves
				builder.Logging.ClearProviders();

				var ip = bindIp == "0.0.0.0" ? IPAddress.Any : IPAddress.Parse(bindIp);

				builder.WebHost.ConfigureKestrel(options =>
				{
					options.Listen(ip, httpPort);
					if (httpsEnabled)
						options.Listen(ip, httpsPort, o => o.UseHttps());
				});

				_app = builder.Build();

				// Multipart upload endpoint
				_app.MapPost("/upload", HandleMultipartUpload);

				// Raw stream upload endpoint
				_app.MapPost("/upload/raw", HandleRawUpload);

				// Single catch-all route
				_app.Map("/{**path}", HandleDefault);

				_listeningUrls.Add($"http://{bindIp}:{httpPort}/");
				if (httpsEnabled)
					_listeningUrls.Add($"https://{bindIp}:{httpsPort}/");

				await _app.StartAsync();

				IsRunning = true;
				ServerEventTriggered?.Invoke(this,
					$"[{DateTime.Now:HH:mm:ss}] [START] Listening on {string.Join(", ", _listeningUrls)}");
				ServerEventTriggered?.Invoke(this,
					$"[{DateTime.Now:HH:mm:ss}] [CONFIG] HTTP port={httpPort} HTTPS={httpsEnabled}");
			}
			catch (Exception ex)
			{
				ServerEventTriggered?.Invoke(this,
					$"[{DateTime.Now:HH:mm:ss}] [ERROR] Failed to start: {ex.Message}");
				await TryCleanupAsync();
			}
		}

		public async Task StopAsync()
		{
			if (!IsRunning) return;
			ServerEventTriggered?.Invoke(this,
				$"[{DateTime.Now:HH:mm:ss}] [STOP REQUEST]");

			await TryCleanupAsync();

			ServerEventTriggered?.Invoke(this,
				$"[{DateTime.Now:HH:mm:ss}] [STOP] Server stopped");
		}

		private async Task TryCleanupAsync()
		{
			try
			{
				if (_app != null)
				{
					await _app.StopAsync();
					await _app.DisposeAsync();
					_app = null;
				}
			}
			catch { }
			finally
			{
				IsRunning = false;
				_listeningUrls.Clear();
			}
		}

		private async Task HandleMultipartUpload(HttpContext context)
		{
			var form = await context.Request.ReadFormAsync();
			var file = form.Files.FirstOrDefault();
			if (file == null)
			{
				context.Response.StatusCode = 400;
				await context.Response.WriteAsync("{\"error\":\"no file\"}");
				return;
			}

			var clientIp = context.Connection.RemoteIpAddress?.ToString() ?? "?";
			ServerEventTriggered?.Invoke(this,
				$"[{DateTime.Now:HH:mm:ss}] [MULTIPART START] {file.FileName} from {clientIp}");

			var saveDir = Path.Combine(AppContext.BaseDirectory, "received_videos");
			Directory.CreateDirectory(saveDir);
			var savePath = Path.Combine(saveDir, file.FileName);
			using (var stream = File.Create(savePath))
				await file.CopyToAsync(stream);

			var receivedAt = DateTime.Now;
			var video = new ReceivedVideo(file.FileName, savePath, receivedAt);
			VideoReceived?.Invoke(this, video);

			context.Response.ContentType = "application/json";
			context.Response.StatusCode = 200;
			await context.Response.WriteAsync("{\"status\":\"uploaded\"}");

			var ts = receivedAt.ToString("HH:mm:ss");
			var sizeKb = file.Length / 1024.0;
			ServerEventTriggered?.Invoke(this,
				$"[{ts}] [UPLOAD] {file.FileName} — {sizeKb:F1} KB — from {clientIp} — saved to {savePath}");
		}

		private async Task HandleRawUpload(HttpContext context)
		{
			var saveDir = Path.Combine(AppContext.BaseDirectory, "received_videos");
			Directory.CreateDirectory(saveDir);

			var fileName = $"video_{DateTime.Now:HHmmss}.mp4";
			var savePath = Path.Combine(saveDir, fileName);

			var expected = context.Request.ContentLength ?? -1;
			foreach (var header in context.Request.Headers)
			{
				ServerEventTriggered?.Invoke(this,
					$"[{DateTime.Now:HH:mm:ss}] [HEADER] {header.Key}: {header.Value}");
			}

			var totalRead = 0L;

			var buffer = new byte[8192];

			var start = DateTime.UtcNow;

			var clientIp = context.Connection.RemoteIpAddress?.ToString() ?? "?";

			ServerEventTriggered?.Invoke(this,
				$"[{DateTime.Now:HH:mm:ss}] [UPLOAD HEADER] Content-Type={context.Request.ContentType} Length={expected}");

			ServerEventTriggered?.Invoke(this,
				$"[{DateTime.Now:HH:mm:ss}] [UPLOAD START] {fileName} from {clientIp} size={expected} bytes");

			using var fs = File.Create(savePath);

			while (true)
			{
				var read = await context.Request.Body.ReadAsync(buffer);

				if (read == 0)
					break;

				await fs.WriteAsync(buffer.AsMemory(0, read));
				totalRead += read;

				if (expected > 0)
				{
					var percent = (double)totalRead / expected * 100.0;

					UploadProgressChanged?.Invoke(this,
						new UploadProgress(totalRead, expected, percent));
				}
				else
				{
					ServerEventTriggered?.Invoke(this,
						$"[{DateTime.Now:HH:mm:ss}] [UPLOAD PROGRESS] {totalRead} bytes");
				}

			}

			var duration = Math.Max((DateTime.UtcNow - start).TotalSeconds, 0.001); // Avoid division by zero
			var rateMbps = (totalRead * 8.0 / 1_000_000.0) / duration;

			ServerEventTriggered?.Invoke(this,
				$"[{DateTime.Now:HH:mm:ss}] [UPLOAD DONE] {fileName} {totalRead} bytes in {duration:F2}s ({rateMbps:F2} Mbps)");

			VideoReceived?.Invoke(this,
				new ReceivedVideo(fileName, savePath, DateTime.Now));

			context.Response.StatusCode = 200;
			await context.Response.WriteAsync("{\"status\":\"uploaded\"}");

		}

		private async Task HandleDefault(HttpContext context)
		{
			var req = context.Request;
			var clientIp = context.Connection.RemoteIpAddress?.ToString() ?? "?";
			var ts = DateTime.Now.ToString("HH:mm:ss");
			var logLine = $"[{ts}] {req.Method} {req.Path}{req.QueryString} from {clientIp}";

			context.Response.ContentType = "application/json";
			context.Response.StatusCode = 200;
			ServerEventTriggered?.Invoke(this,
				$"[{ts}] [REQUEST] {req.Method} {req.Path}{req.QueryString} from {clientIp}");

			await context.Response.WriteAsync("{\"status\":\"ok\",\"server\":\"EmbeddedNetworkLab\"}");

			RequestReceived?.Invoke(this, $"{logLine} → 200");
		}
	}
}

File: backend/src/services/HttpServerService.js
const EventEmitter = require('events');
const express = require('express');
const multer = require('multer');
const http = require('http');
const https = require('https');
const fs = require('fs');
const path = require('path');

const VIDEOS_DIR = path.join(__dirname, '../../../received_videos');

class HttpServerService extends EventEmitter {
  constructor() {
    super();
    this._server = null;
    this.isRunning = false;
    this.listeningUrls = [];
  }

  async start({ bindIp = '0.0.0.0', httpPort = 8081 }) {
    if (this.isRunning) return;

    fs.mkdirSync(VIDEOS_DIR, { recursive: true });

    const app = express();

    // Multer storage: save to received_videos with original filename
    const storage = multer.diskStorage({
      destination: (_req, _file, cb) => cb(null, VIDEOS_DIR),
      filename: (_req, file, cb) => cb(null, file.originalname),
    });
    const upload = multer({ storage });

    // Multipart upload endpoint
    app.post('/upload', upload.single('file'), (req, res) => {
      if (!req.file) {
        return res.status(400).json({ error: 'no file' });
      }
      const clientIp = req.ip;
      const ts = new Date().toLocaleTimeString();
      this.emit('event', `[${ts}] [MULTIPART START] ${req.file.originalname} from ${clientIp}`);

      const video = {
        fileName: req.file.originalname,
        filePath: req.file.path,
        receivedAt: new Date().toISOString(),
      };
      this.emit('video-received', video);

      const sizeKb = (req.file.size / 1024).toFixed(1);
      this.emit('event', `[${ts}] [UPLOAD] ${req.file.originalname} — ${sizeKb} KB — from ${clientIp}`);

      res.json({ status: 'uploaded' });
    });

    // Raw stream upload endpoint
    app.post('/upload/raw', (req, res) => {
      const fileName = `video_${Date.now()}.mp4`;
      const savePath = path.join(VIDEOS_DIR, fileName);
      const expected = parseInt(req.headers['content-length'] || '0', 10);
      const clientIp = req.ip;
      const ts = new Date().toLocaleTimeString();

      this.emit('event', `[${ts}] [UPLOAD START] ${fileName} from ${clientIp} size=${expected} bytes`);

      // Log headers
      for (const [key, val] of Object.entries(req.headers)) {
        this.emit('event', `[${ts}] [HEADER] ${key}: ${val}`);
      }

      const ws = fs.createWriteStream(savePath);
      let totalRead = 0;
      const startMs = Date.now();

      req.on('data', (chunk) => {
        ws.write(chunk);
        totalRead += chunk.length;
        if (expected > 0) {
          const percent = (totalRead / expected) * 100;
          this.emit('upload-progress', { totalRead, expected, percent });
        }
      });

      req.on('end', () => {
        ws.end();
        const duration = Math.max((Date.now() - startMs) / 1000, 0.001);
        const rateMbps = ((totalRead * 8) / 1_000_000 / duration).toFixed(2);
        const doneTs = new Date().toLocaleTimeString();
        this.emit('event', `[${doneTs}] [UPLOAD DONE] ${fileName} ${totalRead} bytes in ${duration.toFixed(2)}s (${rateMbps} Mbps)`);

        this.emit('video-received', {
          fileName,
          filePath: savePath,
          receivedAt: new Date().toISOString(),
        });

        res.json({ status: 'uploaded' });
      });

      req.on('error', (err) => {
        ws.end();
        this.emit('event', `[ERROR] raw upload: ${err.message}`);
        res.status(500).json({ error: err.message });
      });
    });

    // Delete video endpoint
    app.delete('/videos/:fileName', (req, res) => {
      const filePath = path.join(VIDEOS_DIR, path.basename(req.params.fileName));
      const clientIp = req.ip;
      const ts = new Date().toLocaleTimeString();
      if (!fs.existsSync(filePath)) {
        return res.status(404).json({ error: 'Not found' });
      }
      fs.unlinkSync(filePath);
      this.emit('event', `[${ts}] [DELETE] ${req.params.fileName} from ${clientIp}`);
      res.json({ status: 'deleted' });
    });

    // Catch-all: log all requests
    app.use((req, res) => {
      const ts = new Date().toLocaleTimeString();
      this.emit('event', `[${ts}] [REQUEST] ${req.method} ${req.path}${req.url.includes('?') ? '?' + req.url.split('?')[1] : ''} from ${req.ip}`);
      res.json({ status: 'ok', server: 'EmbeddedNetworkLab' });
    });

    return new Promise((resolve, reject) => {
      this._server = http.createServer(app);
      this._server.listen(httpPort, bindIp === '0.0.0.0' ? undefined : bindIp, () => {
        this.isRunning = true;
        this.listeningUrls = [`http://${bindIp}:${httpPort}/`];
        const ts = new Date().toLocaleTimeString();
        this.emit('event', `[${ts}] [START] Listening on ${this.listeningUrls.join(', ')}`);
        resolve();
      });
      this._server.on('error', (err) => {
        this.isRunning = false;
        this.emit('event', `[ERROR] Failed to start: ${err.message}`);
        reject(err);
      });
    });
  }

  async stop() {
    if (!this.isRunning || !this._server) return;
    return new Promise((resolve) => {
      this._server.close(() => {
        this.isRunning = false;
        this._server = null;
        this.listeningUrls = [];
        const ts = new Date().toLocaleTimeString();
        this.emit('event', `[${ts}] [STOP] Server stopped`);
        resolve();
      });
    });
  }

  getReceivedVideos() {
    try {
      if (!fs.existsSync(VIDEOS_DIR)) return [];
      return fs.readdirSync(VIDEOS_DIR)
        .filter(f => /\.(mp4|avi|mkv|mov|h264|bin)$/i.test(f))
        .map(f => ({
          fileName: f,
          filePath: path.join(VIDEOS_DIR, f),
          receivedAt: fs.statSync(path.join(VIDEOS_DIR, f)).mtime.toISOString(),
        }))
        .sort((a, b) => new Date(b.receivedAt) - new Date(a.receivedAt));
    } catch {
      return [];
    }
  }
}

module.exports = HttpServerService;
